core/src/test/java/org/elasticsearch/action/admin/cluster/node/tasks/TransportTasksActionTests.java

```diff
@@ -630,4 +632,69 @@ protected TestTaskResponse taskOperation(TestTasksRequest request, Task task) {
         NodesResponse responses = future.get();
         assertEquals(0, responses.failureCount());
     }
+
+
+    public void testTaskNodeFiltering() throws ExecutionException, InterruptedException, IOException {
+        setupTestNodes(Settings.EMPTY);
+        connectNodes(testNodes);
+        CountDownLatch checkLatch = new CountDownLatch(1);
+        ActionFuture<NodesResponse> future = startBlockingTestNodesAction(checkLatch);
+
+        String[] allNodes = new String[testNodes.length];
+        for (int i = 0; i < testNodes.length; i++) {
+            allNodes[i] = testNodes[i].getNodeId();
+        }
+
+        int filterNodesSize = randomInt(allNodes.length);
+        Set<String> filterNodes = new HashSet<>(randomSubsetOf(filterNodesSize, allNodes));
+        logger.info("Filtering out nodes {} size: {}", filterNodes, filterNodesSize);
+
+        TestTasksAction[] tasksActions = new TestTasksAction[nodesCount];
+        for (int i = 0; i < testNodes.length; i++) {
+            final int node = i;
+            // Simulate task action that fails on one of the tasks on one of the nodes
```

Comment:
The comment is incorrect, it should say "simulate a task action that works on all nodes except nodes listed in filterNodes" or something like this. TestTaskAction is an implementation of TransportTasksAction, which executes on all nodes by default, that's why I need to create N of them. You connect to one of the nodes, but it sends requests to all nodes. It is very similar to TransportNodesAction. Does this make sense? In this test I test that it executes on all nodes except filtered ones. So, I have to create them for all nodes, but you are right I don't have to keep pointers to all of them around but it was easier to implement it this way instead of having a logic to keep a random one.